src/main/java/bio/terra/workspace/service/job/JobService.java

```diff
@@ -0,0 +1,381 @@
+    boolean canListAnyJob = checkUserCanListAnyJob(userReq);
+
+    try {
+      // if the user has access to all jobs, then fetch the requested one
+      // otherwise, check that the user has access to it first
+      if (!canListAnyJob) {
+        verifyUserAccess(jobId, userReq); // jobId=flightId
+      }
+      FlightState flightState = stairway.getFlightState(jobId);
+      return mapFlightStateToJobModel(flightState);
+    } catch (StairwayException stairwayEx) {
+      throw new InternalStairwayException(stairwayEx);
+    }
+  }
+
+  /**
+   * There are four cases to handle here:
+   *
+   * <ol>
+   *   <li>Flight is still running. Throw an JobNotComplete exception
+   *   <li>Successful flight: extract the resultMap RESPONSE as the target class. If a
+   *       statusContainer is present, we try to retrieve the STATUS_CODE from the resultMap and
+   *       store it in the container. That allows flight steps used in async REST API endpoints to
+   *       set alternate success status codes. The status code defaults to OK, if it is not set in
+   *       the resultMap.
+   *   <li>Failed flight: if there is an exception, throw it. Note that we can only throw
+   *       RuntimeExceptions to be handled by the global exception handler. Non-runtime exceptions
+   *       require throw clauses on the controller methods; those are not present in the
+   *       swagger-generated code, so it introduces a mismatch. Instead, in this code if the caught
+   *       exception is not a runtime exception, then we throw JobResponseException passing in the
+   *       Throwable to the exception. In the global exception handler, we retrieve the Throwable
+   *       and use the error text from that in the error model
+   *   <li>Failed flight: no exception present. We throw InvalidResultState exception
+   * </ol>
+   *
+   * @param jobId to process
+   * @return object of the result class pulled from the result map
+   */
+  public <T> JobResultWithStatus<T> retrieveJobResult(
+      String jobId, Class<T> resultClass, AuthenticatedUserRequest userReq) {
+    boolean canListAnyJob = checkUserCanListAnyJob(userReq);
+
+    try {
+      // if the user has access to all jobs, then fetch the requested result
+      // otherwise, check that the user has access to it first
+      if (!canListAnyJob) {
+        verifyUserAccess(jobId, userReq); // jobId=flightId
+      }
+      return retrieveJobResultWorker(jobId, resultClass);
+    } catch (StairwayException stairwayEx) {
+      throw new InternalStairwayException(stairwayEx);
+    }
+  }
+
+  private <T> JobResultWithStatus<T> retrieveJobResultWorker(String jobId, Class<T> resultClass)
+      throws StairwayException {
+    FlightState flightState = stairway.getFlightState(jobId);
+    FlightMap resultMap = flightState.getResultMap().orElse(null);
+    if (resultMap == null) {
+      throw new InvalidResultStateException("No result map returned from flight");
+    }
```

Comment:
I'd prefer this way - I got tripped up by isPresent/null checking earlier on (OpenApi v3 uses JsonNullable instead of Optional, where {is_present:true, value:null} is valid...) and I'd rather not mix similar-looking APIs with different semantics.